Ensure state is decoded before it is processed

jasonnutter opened the pull request:
The library will encode the state value, and in the scenario were there is an error, it appears that the service will encode the state parameter again when it is sent back to the library. This means the library will need to decode it twice before it is in the right format to be parsed. Ensure that the deserialize function decodes twice, and that the parseLibraryState also decodes just in case.

jasonnutter committed:
Ensure state is decoded before it is processed

sameerag reviewed (approved):
LGTM. Was this identified in Unit tests?

pkanher617 commented on lib/msal-core/src/UserAgentApplication.ts:
```diff
@@ -1093,11 +1093,12 @@ export class UserAgentApplication {
             stateResponse = {
                 requestType: Constants.unknown,
-                state: parameters.state,
+                state: decodedState,
```
Any tests we can add for this?

DarylThayil replied:
+1 on tests otherwise lgtm

jasonnutter replied:
Done, please re-review!

jasonnutter committed:
Ensure double encoded values are decoded when parsing hash

jasonnutter commented:
Was able to properly root cause this: we encode the state value on the way out the door, and it appears the service encodes it again when sending back to us. This results in a double encoded state value, so I updated the deserialize function to decode twice to make sure things are in the right format. Added tests.

jmckennon reviewed (approved):
LGTM

jasonnutter commented:
Hmm I seem to be having trouble reproducing this behavior now, maybe it was fixed server side.

jasonnutter committed:
Merge branch 'dev' into decode-library-state
